lighthouse-core/lib/tracehouse/trace-processor.js

```diff
@@ -491,10 +503,140 @@ class TraceProcessor {
     // Filter to just events matching the frame ID, just to make sure.
     const frameEvents = keyEvents.filter(e => e.args.frame === mainFrameIds.frameId);
 
-    // Our navStart will be the last frame navigation in the trace
-    const navigationStart = frameEvents.filter(this._isNavigationStartOfInterest).pop();
-    if (!navigationStart) throw this.createNoNavstartError();
-    const timeOriginEvt = navigationStart;
+    // Compute our time origin to use for all relative timings.
+    const timeOriginEvt = this.computeTimeOrigin(
+      {keyEvents, frameEvents, mainFrameIds},
+      timeOriginDeterminationMethod
+    );
+
+    // Compute the key frame timings for the main frame.
+    const frameTimings = this.computeKeyTimingsForFrame(frameEvents, {timeOriginEvt});
+
+    // Subset all trace events to just our tab's process (incl threads other than main)
+    // stable-sort events to keep them correctly nested.
+    const processEvents = TraceProcessor
+      .filteredTraceSort(trace.traceEvents, e => e.pid === mainFrameIds.pid);
+
+    const mainThreadEvents = processEvents
+      .filter(e => e.tid === mainFrameIds.tid);
+
+    const frames = keyEvents
+      .filter(evt => evt.name === 'FrameCommittedInBrowser')
+      .map(evt => evt.args.data)
+      .filter(/** @return {data is {frame: string, url: string}} */ data => {
+        return Boolean(data && data.frame && data.url);
+      });
+
+    // Ensure our traceEnd reflects all page activity.
+    const traceEnd = this.computeTraceEnd(trace.traceEvents, timeOriginEvt);
```

Comment:
any reason not to just pass `trace.traceEvents` into `computeKeyTimingsForFrame` and do `traceEnd` for real in there?

Reply:
As in pass both frameEvents and all trace events?

Primary reason I did not do that is it violates the veil of abstraction that `computeKeyTimingsForFrame` well, computes the key timings for a frame :) 

If the double computation is problematic I'd prefer to modify the return type of `computeKeyTimingsForFrame` to omit `traceEnd` entirely as a proper fix. The duplication of all the types to avoid a double computation didn't seem to be worth the tradeoff in confusion initially since the trace end computation is cheap (~0.5ms/MB of trace). Prefer this approach?

Reply:
> If the double computation is problematic

It's not fatally problematic, no more so than breaking the veil :) And definitely not worried about the computation cost. I was just curious about your thinking on which was less confusing, passing in both sets of events or fake computing it and then doing it for real. Breaking the type down might be worse but it's possible it would be ok...e.g. with an inferred return type on `computeKeyTimingsForFrame`, maybe?

I'm also trying to be more literal in my review comments. I was 95% sure you had already considered this question so I was 95% really just asking, and 5% suggesting considering it and then just asking as a consideration follow up :)

Reply:
> Breaking the type down might be worse but it's possible it would be ok...e.g. with an inferred return type on computeKeyTimingsForFrame, maybe?

I like it SG!

Reply:
> If the double computation is problematic I'd prefer to modify the return type of computeKeyTimingsForFrame to omit traceEnd entirely as a proper fix.

Just noticed `FrameTimings` is a new typedef, and yeah, this would be the proper fix and it would be great if it's a simple modification to the `Omit`, but I don't know if there's an easy way to do that one level deep, so don't worry about it if there's not.